File: 2019-01-16/11041/11041.java
import java.io.*;

class Main{
	static long P = 1000000007;

	public static void main(String[] args)throws Exception{
		BufferedReader bfr = new BufferedReader(new InputStreamReader(System.in));
		String[] NK = bfr.readLine().split(" ");

		long n = strToLong(NK[0]);
		long k = strToLong(NK[1]);
		long result;

		long A = factorial(1,n);
		long B = factorial(1,k);
		B = factorial(B, n-k);

		B = calculateDegree(B, P-2);
		B%=P;
		result = A*B;
		result%=P;
		System.out.println(result);
	}

	public static long strToLong(String str){
		return Long.parseLong(str);
	}

	public static long factorial(long operator, long limit){
		for(long i = 1; i <= limit; i++){
			operator*=i;
			operator%=P;
		}
		return operator;
	}

	public static long calculateDegree(long operator, long limit){
		long ans = 1;
		while(limit>0){
			if(limit%2 != 0){
				ans*=operator;
				ans%=P;
			}
			operator*=operator;
			operator%=P;
			limit/=2;
		}
		return ans;	
	}	
}

File: 2019-01-16/11041/11041.c
#include <stdio.h>

#define P 1000000007

long long factorial(long long operator, long long limit);
long long mulMod(long long operatorA, long long operatorB);
long long calculOrder(long long operator, long long limit);

int main(){
	long long n, k;
	long long result;
	scanf("%lld %lld", &n, &k);
	long long A = factorial(1, n);
	long long B = factorial(1, k);
	B = factorial(B, n-k);
	B = calculOrder(B, P-2);

	B%=P;
	result = A*B;
	result%=P;
	printf("%lld", result);
}

long long factorial(long long operator, long long limit){
	for(long long i = 1; i <= limit;i++)
		operator = mulMod(operator, i);
	
	return operator;
}

long long calculOrder(long long operator, long long limit){
	long long result = 1;

	while(limit>0){
		if(limit%2!=0)
			result = mulMod(result, operator);
		
		operator = mulMod(operator, operator);
		limit/=2;
	}
	return result;
}

long long mulMod(long long operatorA, long long operatorB){
	operatorA*=operatorB;
	return operatorA%P;
}